--- postgresql/src/include/port/emscripten.h
+++ postgresql-wasm/src/include/port/emscripten.h
@@ -0,0 +1,18 @@
+/* src/include/port/emscripten.h */
+
+#ifndef I_EMSCRIPTEN
+#define I_EMSCRIPTEN
+
+#if !defined(__cplusplus)
+#include <emscripten.h>
+#endif
+
+#include "/tmp/pglite/include/wasm_common.h"
+
+#include <stdbool.h>
+extern PGDLLIMPORT bool fe_utils_quote_all_identifiers;
+
+#define BOOT_END_MARK "build indices"
+#define FD_BUFFER_MAX 16384
+
+#endif // I_EMSCRIPTEN
